File: api/src/models/workout_log.py
from pydantic import BaseModel


class CreateWorkoutLogRequest(BaseModel):
    routine_id: int
    date: str
    notes: str | None = None


class LogSetRequest(BaseModel):
    exercise_id: int
    set_number: int
    reps: int
    weight: float | None = None


class UpdateWorkoutLogRequest(BaseModel):
    date: str | None = None
    notes: str | None = None
    completed: bool | None = None


class UpdateSetRequest(BaseModel):
    reps: int | None = None
    weight: float | None = None


class SetLogResponse(BaseModel):
    id: int | None = None
    workout_log_id: int
    exercise_id: int
    set_number: int
    reps: int | None = None
    weight: float | None = None
    created_at: str | None = None
    exercise_name: str | None = None


class SetHistoryEntry(BaseModel):
    set_number: int
    reps: int | None = None
    weight: float | None = None
    workout_log_id: int
    date: str
    routine_name: str | None = None


class WorkoutLogResponse(BaseModel):
    id: int
    routine_id: int
    routine_name: str | None = None
    date: str
    notes: str | None = None
    created_at: str | None = None
    completed: bool
    sets: list[SetLogResponse] = []


class RoutineLogSummaryResponse(BaseModel):
    """One past session against a routine, with aggregates for a history view."""

    id: int
    date: str
    notes: str | None = None
    created_at: str | None = None
    completed: bool
    total_sets: int = 0
    total_volume: float = 0

File: ui/src/types/WorkoutLog.ts
export interface WorkoutLog {
    id: number;
    routine_id: number;
    routine_name: string;
    date: string;
    notes: string | null;
    created_at: string;
    completed: boolean;
}

export interface SetLog {
    id: number;
    workout_log_id: number;
    exercise_id: number;
    set_number: number;
    reps: number;
    weight: number | null;
    exercise_name: string;
}

export interface ExerciseHistoryEntry {
    workout_log_id: number;
    date: string;
    routine_name: string;
    set_number: number;
    reps: number;
    weight: number | null;
}

export interface SetLogUpdate {
    reps?: number | null;
    weight?: number | null;
}

export interface WorkoutLogUpdate {
    date?: string | null;
    notes?: string | null;
    completed?: boolean | null;
}

export interface WorkoutLogDetail {
    id: number;
    routine_id: number;
    routine_name: string | null;
    date: string;
    notes: string | null;
    created_at: string;
    completed: boolean;
    sets: SetLog[];
}

export interface RoutineLogSummary {
    id: number;
    date: string;
    notes: string | null;
    created_at: string | null;
    completed: boolean;
    total_sets: number;
    total_volume: number;
}
